celerpy/visualize.py

```diff
@@ -0,0 +1,352 @@
+
+    def trace(
+        self,
+        image: Optional[model.ImageInput] = None,
+        *,
+        geometry: Optional[model.GeometryEngine] = None,
+        **kwargs,
+    ):
+        """Trace with a geometry, memspace, etc."""
+        if image is None and not self.image:
+            raise RuntimeError(
+                "Image specifications must be supplied for the first trace"
+            )
+
+        # Determine whether to request the list of volumes for this geometry
+        volumes = None
+        if geometry is not None:
+            volumes = self.volumes.setdefault(geometry, [])
+
+        with NamedTemporaryFile(suffix=".bin", mode="w+b") as f:
+            inp = model.TraceInput(
+                geometry=geometry,
+                volumes=(not volumes),
+                bin_file=Path(f.name),
+                image=image,
+                **kwargs,
+            )
+            result = process.communicate_model(
+                self.process, inp, model.TraceOutput
+            )
+            img = f.read()
+
+        # Cache the geometry names and ensure trace has them
+        if geometry is None:
+            geometry = result.trace.geometry
+            assert geometry is not None
+            volumes = self.volumes.setdefault(geometry, [])
+        if not volumes:
+            assert isinstance(volumes, list)
+            assert result.volumes
+            # XXX : erase pointer names (for now?)
+            volumes[:] = [_re_ptr.sub("", v) for v in result.volumes]
+        else:
+            result.volumes = volumes
+
+        # Reinterpret the image as a correctly shaped Numpy array
+        npimg = np.frombuffer(img, dtype=np.int32).reshape(result.image.dims)
+        self.image = result.image
+
+        return (result, npimg)
+
+    def close(self, *, timeout=0.25):
+        """Cleanly exit the ray trace loop, returning run statistics if possible."""
+        result = process.communicate(self.process, json.dumps(None))
+        try:
+            self.process.wait(timeout=timeout)
+        except TimeoutExpired:
+            pass
+        result = result or process.close(self.process, timeout=timeout)
+        try:
+            result = json.loads(result)
+        except json.JSONDecodeError:
+            pass
+        return result
+
+
+class ReverseIndexDict(collections.defaultdict):
+    """Manage a two-way mapping of integers and another type."""
+
+    def __init__(self, volume_list):
+        self.vlist = volume_list
+
+    def __missing__(self, key):
+        self[key] = result = len(self)
+        self.vlist.append(key)
+        assert len(self.vlist) == len(self)
+        return result
+
+
+LabeledAxis = collections.namedtuple("LabeledAxis", ["label", "lo", "hi"])
+LabeledAxes = collections.namedtuple("LabeledAxes", ["x", "y"])
+
+
+class IdNorm(BoundaryNorm):
+    """Map IDs to consecutive integers for rendering.
+
+    This provides literally a factor of 1000 speedup over using BoundaryNorm
```

Comment:
:exploding_head:

Reply:
It did at the time I wrote it 5 years ago 😅 maybe they've improved it since then